vector/src/main/java/im/vector/app/features/widgets/ptt/BluetoothLowEnergyService.kt

```diff
@@ -0,0 +1,125 @@
+import android.bluetooth.BluetoothGatt
+import android.bluetooth.BluetoothGattCallback
+import android.bluetooth.BluetoothGattCharacteristic
+import android.bluetooth.BluetoothManager
+import android.bluetooth.BluetoothProfile
+import android.content.Intent
+import android.os.Binder
+import android.os.Build
+import android.os.IBinder
+import im.vector.app.core.services.VectorService
+import androidx.core.content.getSystemService
+import dagger.hilt.android.AndroidEntryPoint
+import im.vector.app.features.notifications.NotificationUtils
+import timber.log.Timber
+import javax.inject.Inject
+import kotlin.random.Random
+
+@AndroidEntryPoint
+class BluetoothLowEnergyService : VectorService() {
+
+    interface Callback {
+        fun onCharacteristicRead(data: ByteArray)
+    }
+
+    @Inject lateinit var notificationUtils: NotificationUtils
+
+    private var bluetoothAdapter: BluetoothAdapter? = null
+    private var bluetoothGatt: BluetoothGatt? = null
+
+    private val binder = LocalBinder()
+
+    var callback: Callback? = null
+
+    private val gattCallback = object : BluetoothGattCallback() {
+        override fun onConnectionStateChange(gatt: BluetoothGatt?, status: Int, newState: Int) {
+            when (newState) {
+                BluetoothProfile.STATE_CONNECTING -> Timber.d("### BluetoothLowEnergyService.newState: STATE_CONNECTING")
+                BluetoothProfile.STATE_CONNECTED -> {
+                    Timber.d("### BluetoothLowEnergyService.newState: STATE_CONNECTED")
+                    bluetoothGatt?.discoverServices()
+                }
+                BluetoothProfile.STATE_DISCONNECTING -> Timber.d("### BluetoothLowEnergyService.newState: STATE_DISCONNECTING")
+                BluetoothProfile.STATE_DISCONNECTED -> Timber.d("### BluetoothLowEnergyService.newState: STATE_DISCONNECTED")
+            }
+        }
+
+        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
+            gatt.services.forEach { service ->
+                service.characteristics.forEach { characteristic ->
+                    gatt.setCharacteristicNotification(characteristic, true)
+                }
+            }
+        }
+
+        override fun onCharacteristicRead(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, status: Int) {
+            if (status == BluetoothGatt.GATT_SUCCESS) {
+                onCharacteristicRead(characteristic)
+            }
+        }
+
+        override fun onCharacteristicChanged(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
+            onCharacteristicRead(characteristic)
+        }
+    }
+
+    override fun onCreate() {
+        super.onCreate()
+        initializeBluetoothAdapter()
+    }
+
+    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+        val notification = notificationUtils.buildBluetoothLowEnergyNotification()
+        startForeground(Random.nextInt(), notification)
```

Comment:
We need to kill the service when the WidgetFragment is destroyed.

Reply:
We should also disconnect from the bluetooth device which it looks like we might not be doing currently